modyn/trainer_server/internal/dataset/online_dataset.py

```diff
@@ -138,23 +149,51 @@ def _info(self, msg: str, worker_id: Optional[int]) -> None:  # pragma: no cover
     def _debug(self, msg: str, worker_id: Optional[int]) -> None:  # pragma: no cover
         logger.debug(f"[Training {self._training_id}][PL {self._pipeline_id}][Worker {worker_id}] {msg}")
 
-    def _get_data(self, worker_id: int, partition_id: int) -> tuple[list[int], list[bytes], list[int]]:
-        self._info("Getting keys from selector", worker_id)
-        keys = self._get_keys_from_selector(worker_id, partition_id)
+    def switch_to_local_key_source(self) -> None:
+        self._keys_source = "local"
+        self._local_dataset_handler = LocalDatasetHandler(self._pipeline_id, self._trigger_id, self._number_of_workers)
+
+    def switch_to_selector_key_source(self) -> None:
+        if self._keys_source == "selector":
+            return
+        self._keys_source = "selector"
+        assert self._local_dataset_handler is not None
+        self._local_dataset_handler.clean_working_directory()
+        self._local_dataset_handler = None
+
+    def _get_data(self, worker_id: int, partition_id: int) -> tuple[list[int], list[bytes], list[int], list[float]]:
+        # Keys can either be retreived from the selector or from local.
+        assert self._keys_source in ["local", "selector"]
+
+        if self._keys_source == "selector":
+            self._info("Getting keys from selector", worker_id)
+            keys = self._get_keys_from_selector(worker_id, partition_id)
+            # right now the selector does not store weights, so a list of ones is returned
+            weights = [1.0] * len(keys)
```

Comment:
the selector does not store weights? i think it does store them, right? _get_keys_from_selector just does not return them? i think we do store and send them?

Reply:
I had some comments on how weights are currently used now in the other thread. Just to make it clear: Weights can come, e.g., by marking specific data points as important, e.g., making a customer important, or saying pictures from this car at night might get double the impact. Right now I think the weight changes caused many if/elses and I was not sure why we only care for the weights in StB. in general, the weights are a "modifier" to the gradient, right, like a multiplicator, so it should be independent of downsampling?